main.go

```diff
@@ -87,39 +90,41 @@ func (*noCopy) Unlock() {}
 // Server represents a PPOPRF randomness server instance.
 type Server struct {
 	sync.Mutex
-	raw    *C.RandomnessServer
-	noCopy noCopy //nolint:structcheck
-	md     epoch
-	done   chan bool
-	pubKey string // Base64-encoded public key.
+	raw            *C.RandomnessServer
+	noCopy         noCopy //nolint:structcheck
+	done           chan bool
+	pubKey         string // Base64-encoded public key.
+	firstEpochTime time.Time
+	epochLen       time.Duration
 }
 
 // epochLoop periodically punctures the randomness server's PPOPRF and -- if
 // necessary -- re-creates the randomness server instance.
-func (srv *Server) epochLoop(epochLen time.Duration) {
+func (srv *Server) epochLoop() {
 	// Odds are that the server's start time does not coincide with the next
 	// epoch's start time.  We therefore wait until the next epoch begins, at
 	// which point we begin our epoch rotation loop.
 	now := time.Now().UTC()
-	_, nextEpochTs := getEpoch(epochLen, firstEpochTs, now)
-	diff := nextEpochTs.Sub(now)
-	elog.Printf("Waiting %s until next epoch begins at %s.", diff, nextEpochTs)
+	currentEpoch, nextEpochTime := srv.getEpoch(now)
+	diff := nextEpochTime.Sub(now)
+	elog.Printf("Waiting %s until next epoch begins at %s.", diff, nextEpochTime)
 	<-time.NewTicker(diff).C
 
-	ticker := time.NewTicker(epochLen)
+	ticker := time.NewTicker(srv.epochLen)
 	elog.Println("Starting epoch loop.")
 	for {
 		select {
 		case <-srv.done:
 			return
 		case <-ticker.C:
-			if err := srv.puncture(); err != nil {
+			if err := srv.puncture(currentEpoch); err != nil {
 				if err.Error() == errEpochExhausted {
 					if err := srv.init(); err != nil {
 						elog.Fatal("Failed to re-initialize randomness server.")
 					}
 				}
 			}
+			currentEpoch++
```

Comment:
getServerInfo calls getEpoch for each request, but this code calls it once at startup and assumes epochs advance uniformly from there. Might be better to call getEpoch each time through the loop so control of the epoch boundaries is in one place.